src/PreSmalltalks-Pharo/Object.extension.st

```diff
@@ -1,7 +1,35 @@
 Extension { #name : #Object }
 
+{ #category : #'*PreSmalltalks-Pharo' }
+Object >> perform: aSymbol [ 
+	^ self perform: aSymbol withArguments: #()
+]
+
+{ #category : #'*PreSmalltalks-Pharo' }
+Object >> perform: selector withArguments: argArray [ 
+	^ self perform: selector withArguments: argArray inSuperclass: self class
+]
+
 { #category : #'*PreSmalltalks-Pharo' }
 Object >> perform: selector withArguments: argArray inSuperclass: lookupClass [
+	selector numArgs = argArray size
+		ifTrue: [ ^self primPerform: selector withArguments: argArray inSuperclass: lookupClass ].
+
+	^selector numArgs < argArray size
+		ifTrue: [ self error: 'cannot uncurry' ]
+		ifFalse: [ "curry"
+			| n |
+			n := selector numArgs - argArray size.
+			{
+			[ :x       | self perform: selector withArguments: argArray, {x} inSuperclass: lookupClass ].
+			[ :x :y    | self perform: selector withArguments: argArray, {x.y} inSuperclass: lookupClass ].
+			[ :x :y :z | self perform: selector withArguments: argArray, {x.y.z} inSuperclass: lookupClass ].
+			} at: n
+		].
+]
+
```

Comment:
This implementation IMO has number of problems: 
1) what if `selector` is not a symbol? This was checked here and is checked in `primPerform.`
2) `numArgs` is potentially sent twice and `numArgs` is rather slow
3) What if we want pass 4 arguments? Then we get index out of bounds, not very friendly. 

Also. now `obj perform: #x:y: with: a` does something else than `obj perform:#x:y: withArguments: { a }` and nowhere in the code is any kind of warning.

Reply:
> what if selector is not a symbol?

Generalizing method lookup on the left of `#at`: is wholly future work.  We have done a little bit on the right of `#at:`, in order to understand what happens when in

```
A   B
-----
  C
```
the `A   B ` part is not a `CompiledMethod` (the Back-to-the-Future Squeak handles it just fine although Calypso tooling crashes in many places), but how to generalize `C`-matching from symbol lookup to general unification is not clear.

I *think* giving the VM prim a chance to unify `C` is what we want here, so we should NOT check for `isSymbol`.  If  `selector` does not understand `numArgs`, that case is handled in `Object>>doesNotUnderstand:` in the most perfect way I can think of.

> This was checked here

No, it was part of recovery after VM prim failure (which never did anything meaningful beyond raising a useless uninformative error anyway), NOT before attempting VM prim.

>  and is checked in primPerform

No, same here: it was part of recovery after VM prim failure (which never did anything meaningful beyond raising a useless uninformative error anyway), NOT before attempting VM prim.

> numArgs is potentially sent twice and numArgs is rather slow

Ok, what you are suggesting is to cache numArgs in a temp?  I can do that.

> What if we want pass 4 arguments?

Yeah, this is mirrorning the current state of block currying — that one also crashes the same way if currying more than 3 args.  I shall fix this some day.

> now obj perform: #x:y: with: a does something else than obj perform:#x:y: withArguments: { a } and nowhere in the code is any kind of warning

Yes.  This is exactly the problem I meant when I said "_This makes me really unhappy, because we'd have to special-case these, and this is really horrible._"  Perhaps I can put a "Caveat" note in that method, in lieu of a real working solution.  This whole PR is a workaround for a VM bug: monkey hackery trying to avoid the situation where the VM corrupts the stack.

Reply:
> 
> I _think_ giving the VM prim a chance to unify `C` is what we want here, so we should NOT check for `isSymbol`. If `selector` does not understand `numArgs`, that case is handled in `Object>>doesNotUnderstand:` in the most perfect way I can think of.

I disagree. This is about giving user a clue what went wrong (both, what really went wrong and under what assumption whoever wrote the code wrote it - whether they're useful or not, correct or not, is different story). 

Besides, if you really want to relax and support other objects than symbols as selectors, the code is bogus anyway, because it is not the selector per se you should ask for `numArgs`, it is the method that's going to be executed. 

> 
> > This was checked here
> 
> No, it was part of recovery after VM prim failure (which never did anything meaningful beyond raising a useless uninformative error anyway), NOT before attempting VM prim.
> 
> > and is checked in primPerform
> 
> No, same here: it was part of recovery after VM prim failure (which never did anything meaningful beyond raising a useless uninformative error anyway), NOT before attempting VM prim.
> 
Again, I disagree with the part "useless uninformative error".

> > numArgs is potentially sent twice and numArgs is rather slow
> 
> Ok, what you are suggesting is to cache numArgs in a temp? I can do that.

Yes, thanks!

> 
> > What if we want pass 4 arguments?
> 
> Yeah, this is mirrorning the current state of block currying — that one also crashes the same way if currying more than 3 args. I shall fix this some day.

I must admit I have little trust in "fix it some day". That day will come, surely. At the least convenient time. 
> 
> > now obj perform: #x:y: with: a does something else than obj perform:#x:y: withArguments: { a } and nowhere in the code is any kind of warning
> 
> ... Perhaps I can put a "Caveat" note in that method, in lieu of a real working solution...

I'd like that, yes. 

As a side note, I spent few hours yesterday looking into fixing the VM properly but the stuff is really messed up.

Reply:
> cache numArgs in a temp

Done.

> put a "Caveat" note in that method, in lieu of a real working solution

Done.

> This is about giving user a clue what went wrong

Right, so we are continuing our decade-long conversation about the scientific approach to Building the Steam Engine.  This `error:` is trying to explicate programmer intent, except that `Object>>error:` is broken (wrongly assumes flat spacetime) so we shall remove it from the system, and replace it with `assume:`/`assert:` when we enough Smalltalk-25 to be able to do it.  We are on the same page here.